src/box/lua/config/instance_config.lua

```diff
@@ -224,20 +225,28 @@ end
 --
 -- See the uri_is_suitable_to_connect() method in the instance
 -- schema object for details.
-local function find_suitable_uri_to_connect(listen)
+local function find_suitable_uri_to_connect(listen, opts)
+    local opts = opts or {}
+    assert(opts.log_prefix == nil or type(opts.log_prefix) == 'string')
+
     for _, u in ipairs(listen) do
         assert(u.uri ~= nil)
         local uri = urilib.parse(u.uri)
-        if uri ~= nil and uri_is_suitable_to_connect(uri) then
-            -- The urilib.format() call has the second optional
-            -- argument `write_password`. Let's assume that the
-            -- given URIs are to listen on them and so have no
-            -- user/password.
-            --
-            -- NB: We need to format the URI back to construct the
-            -- 'uri' field. urilib.parse() creates separate 'host'
-            -- and 'service'.
-            return urilib.format(uri), u.params
+        if uri ~= nil then
+            local ok, err = uri_is_suitable_to_connect(uri)
+            if ok then
+                -- The urilib.format() call has the second optional
+                -- argument `write_password`. Let's assume that the
+                -- given URIs are to listen on them and so have no
+                -- user/password.
+                --
+                -- NB: We need to format the URI back to construct the
+                -- 'uri' field. urilib.parse() creates separate 'host'
+                -- and 'service'.
+                return urilib.format(uri), u.params
+            elseif opts.log_prefix then
+                log.warn(opts.log_prefix .. err)
```

Comment:
The URI itself is not shown in the log.

<details>
<summary>config.yaml</summary>

```yaml
credentials:
  users:
    replicator:
      password: 'secret'
      roles: [replication]
    client:
      password: 'secret'
      roles: [super]

iproto:
  advertise:
    peer:
      login: 'replicator'

replication:
  failover: election

groups:
  group-001:
    replicasets:
      replicaset-001:
        instances:
          instance-001:
            iproto:
              listen:
                - uri: localhost:0
                - uri: 0.0.0.0:33301
                - uri: localhost:3301
          instance-002:
            iproto:
              listen:
                - uri: localhost:0
                - uri: 0.0.0.0:33302
                - uri: localhost:3302
          instance-003:
            iproto:
              listen:
                - uri: localhost:0
                - uri: 0.0.0.0:33303
                - uri: localhost:3303
```
</details>

<details>
<summary>Procfile (for honcho)</summary>

```procfile
instance-001: ./src/tarantool --config config.yaml --name instance-001
instance-002: ./src/tarantool --config config.yaml --name instance-002
instance-003: ./src/tarantool --config config.yaml --name instance-003
```
</details>

<details>
<summary>Logs</summary>

```sh
14:55:11 instance-002.1 | replicaset dataflow configuration: An URI with zero port cannot be used to create a client socket
14:55:11 instance-002.1 | replicaset dataflow configuration: INADDR_ANY (0.0.0.0) cannot be used to create a client socket
14:55:11 instance-001.1 | replicaset dataflow configuration: An URI with zero port cannot be used to create a client socket
14:55:11 instance-001.1 | replicaset dataflow configuration: INADDR_ANY (0.0.0.0) cannot be used to create a client socket
14:55:11 instance-001.1 | replicaset dataflow configuration: An URI with zero port cannot be used to create a client socket
14:55:11 instance-001.1 | replicaset dataflow configuration: INADDR_ANY (0.0.0.0) cannot be used to create a client socket
14:55:11 instance-001.1 | replicaset dataflow configuration: An URI with zero port cannot be used to create a client socket
14:55:11 instance-002.1 | replicaset dataflow configuration: An URI with zero port cannot be used to create a client socket
14:55:11 instance-001.1 | replicaset dataflow configuration: INADDR_ANY (0.0.0.0) cannot be used to create a client socket
14:55:11 instance-002.1 | replicaset dataflow configuration: INADDR_ANY (0.0.0.0) cannot be used to create a client socket
14:55:11 instance-003.1 | replicaset dataflow configuration: An URI with zero port cannot be used to create a client socket
14:55:11 instance-002.1 | replicaset dataflow configuration: An URI with zero port cannot be used to create a client socket
14:55:11 instance-003.1 | replicaset dataflow configuration: INADDR_ANY (0.0.0.0) cannot be used to create a client socket
14:55:11 instance-002.1 | replicaset dataflow configuration: INADDR_ANY (0.0.0.0) cannot be used to create a client socket
14:55:11 instance-003.1 | replicaset dataflow configuration: An URI with zero port cannot be used to create a client socket
14:55:11 instance-003.1 | replicaset dataflow configuration: INADDR_ANY (0.0.0.0) cannot be used to create a client socket
14:55:11 instance-003.1 | replicaset dataflow configuration: An URI with zero port cannot be used to create a client socket
14:55:11 instance-003.1 | replicaset dataflow configuration: INADDR_ANY (0.0.0.0) cannot be used to create a client socket
```
</details>

Reply:
An example with a sharding configuration.


<details>
<summary>config.yaml</summary>

```yaml
credentials:
  users:
    replicator:
      password: 'secret'
      roles: [replication]
    storage:
      password: 'secret'
      roles: [sharding]
    client:
      password: 'secret'
      roles: [super]

iproto:
  advertise:
    peer:
      login: 'replicator'
    sharding:
      login: 'storage'

replication:
  failover: election

groups:
  storage-group-001:
    sharding:
      roles: [storage]
    replicasets:
      storage-rs-001:
        instances:
          storage-rs-001-i-001:
            iproto:
              listen:
                - uri: localhost:0
                - uri: 0.0.0.0:33311
                - uri: localhost:3311
          storage-rs-001-i-002:
            iproto:
              listen:
                - uri: localhost:0
                - uri: 0.0.0.0:33312
                - uri: localhost:3312
          storage-rs-001-i-003:
            iproto:
              listen:
                - uri: localhost:0
                - uri: 0.0.0.0:33313
                - uri: localhost:3313
      storage-rs-002:
        instances:
          storage-rs-002-i-001:
            iproto:
              listen:
                - uri: localhost:0
                - uri: 0.0.0.0:33321
                - uri: localhost:3321
          storage-rs-002-i-002:
            iproto:
              listen:
                - uri: localhost:0
                - uri: 0.0.0.0:33322
                - uri: localhost:3322
          storage-rs-002-i-003:
            iproto:
              listen:
                - uri: localhost:0
                - uri: 0.0.0.0:33333
                - uri: localhost:3333
  router-group-001:
    sharding:
      roles: [router]
    replicasets:
      router-rs-001:
        instances:
          router-001: {}
      router-rs-002:
        instances:
          router-002: {}
      router-rs-003:
        instances:
          router-003: {}
```
</details>


<details>
<summary>Procfile</summary>

```procfile
storage-rs-001-i-001: ./src/tarantool --config config.yaml --name storage-rs-001-i-001
storage-rs-001-i-002: ./src/tarantool --config config.yaml --name storage-rs-001-i-002
storage-rs-001-i-003: ./src/tarantool --config config.yaml --name storage-rs-001-i-003
storage-rs-002-i-001: ./src/tarantool --config config.yaml --name storage-rs-002-i-001
storage-rs-002-i-002: ./src/tarantool --config config.yaml --name storage-rs-002-i-002
storage-rs-002-i-003: ./src/tarantool --config config.yaml --name storage-rs-002-i-003
router-001: ./src/tarantool --config config.yaml --name router-001
router-002: ./src/tarantool --config config.yaml --name router-002
router-003: ./src/tarantool --config config.yaml --name router-003
```
</details>

<details>
<summary>Logs</summary>

```
15:18:51 router-001.1           | 2024-04-14 15:18:51.397 [20675] main/104/interactive/tarantool.config log.lua:130 W> sharding configuration: An URI with zero port cannot be used to create a client socket
15:18:51 router-001.1           | 2024-04-14 15:18:51.397 [20675] main/104/interactive/tarantool.config log.lua:130 W> sharding configuration: INADDR_ANY (0.0.0.0) cannot be used to create a client socket
15:18:51 router-001.1           | 2024-04-14 15:18:51.398 [20675] main/104/interactive/tarantool.config log.lua:130 W> sharding configuration: An URI with zero port cannot be used to create a client socket
15:18:51 router-001.1           | 2024-04-14 15:18:51.398 [20675] main/104/interactive/tarantool.config log.lua:130 W> sharding configuration: INADDR_ANY (0.0.0.0) cannot be used to create a client socket
15:18:51 router-001.1           | 2024-04-14 15:18:51.398 [20675] main/104/interactive/tarantool.config log.lua:130 W> sharding configuration: An URI with zero port cannot be used to create a client socket
15:18:51 router-001.1           | 2024-04-14 15:18:51.398 [20675] main/104/interactive/tarantool.config log.lua:130 W> sharding configuration: INADDR_ANY (0.0.0.0) cannot be used to create a client socket
15:18:51 router-001.1           | 2024-04-14 15:18:51.400 [20675] main/104/interactive/tarantool.config log.lua:130 W> sharding configuration: An URI with zero port cannot be used to create a client socket
15:18:51 router-001.1           | 2024-04-14 15:18:51.400 [20675] main/104/interactive/tarantool.config log.lua:130 W> sharding configuration: INADDR_ANY (0.0.0.0) cannot be used to create a client socket
15:18:51 router-001.1           | 2024-04-14 15:18:51.400 [20675] main/104/interactive/tarantool.config log.lua:130 W> sharding configuration: An URI with zero port cannot be used to create a client socket
15:18:51 router-001.1           | 2024-04-14 15:18:51.400 [20675] main/104/interactive/tarantool.config log.lua:130 W> sharding configuration: INADDR_ANY (0.0.0.0) cannot be used to create a client socket
15:18:51 router-001.1           | 2024-04-14 15:18:51.401 [20675] main/104/interactive/tarantool.config log.lua:130 W> sharding configuration: An URI with zero port cannot be used to create a client socket
15:18:51 router-001.1           | 2024-04-14 15:18:51.401 [20675] main/104/interactive/tarantool.config log.lua:130 W> sharding configuration: INADDR_ANY (0.0.0.0) cannot be used to create a client socket
```
</details>

Reply:
Well, yea, this is a bit inconsistent considering the previous commit. I added the same "bad URI %q:" here.

<details>
<summary>Replicaset logs:</summary>

```
20:17:48 instance-001.1 | replicaset dataflow configuration: bad URI "localhost:0" provided: An URI with zero port cannot be used to create a client socket
20:17:48 instance-001.1 | replicaset dataflow configuration: bad URI "0.0.0.0:33301" provided: INADDR_ANY (0.0.0.0) cannot be used to create a client socket
20:17:48 instance-001.1 | replicaset dataflow configuration: bad URI "localhost:0" provided: An URI with zero port cannot be used to create a client socket
20:17:48 instance-001.1 | replicaset dataflow configuration: bad URI "0.0.0.0:33302" provided: INADDR_ANY (0.0.0.0) cannot be used to create a client socket
20:17:48 instance-001.1 | replicaset dataflow configuration: bad URI "localhost:0" provided: An URI with zero port cannot be used to create a client socket
20:17:48 instance-001.1 | replicaset dataflow configuration: bad URI "0.0.0.0:33303" provided: INADDR_ANY (0.0.0.0) cannot be used to create a client socket
```
</details>

<details>
<summary>Sharding logs:</summary>

```
20:20:58 storage-rs-001-i-003.1 | 2024-04-14 20:20:58.885 [298006] main/104/interactive/tarantool.config log.lua:130 W> sharding configuration: bad URI "localhost:0" provided: An URI with zero port cannot be used to create a client socket
20:20:58 storage-rs-001-i-003.1 | 2024-04-14 20:20:58.885 [298006] main/104/interactive/tarantool.config log.lua:130 W> sharding configuration: bad URI "0.0.0.0:33311" provided: INADDR_ANY (0.0.0.0) cannot be used to create a client socket
20:20:58 storage-rs-001-i-003.1 | 2024-04-14 20:20:58.886 [298006] main/104/interactive/tarantool.config log.lua:130 W> sharding configuration: bad URI "localhost:0" provided: An URI with zero port cannot be used to create a client socket
20:20:58 storage-rs-001-i-003.1 | 2024-04-14 20:20:58.887 [298006] main/104/interactive/tarantool.config log.lua:130 W> sharding configuration: bad URI "0.0.0.0:33312" provided: INADDR_ANY (0.0.0.0) cannot be used to create a client socket
20:20:58 storage-rs-001-i-003.1 | 2024-04-14 20:20:58.887 [298006] main/104/interactive/tarantool.config log.lua:130 W> sharding configuration: bad URI "localhost:0" provided: An URI with zero port cannot be used to create a client socket
20:20:58 storage-rs-001-i-003.1 | 2024-04-14 20:20:58.887 [298006] main/104/interactive/tarantool.config log.lua:130 W> sharding configuration: bad URI "0.0.0.0:33313" provided: INADDR_ANY (0.0.0.0) cannot be used to create a client socket
```
</details>

Reply:
I am a bit concerned that the logs are getting too long, but I don't see much we can do about it besides removing "provided" or any other supplementary word.

Reply:
Nice, thanks!

Is there a reason why we don't add this `bad URI %q provided` clause right into `uri_is_suitable_to_connect()`? This way we have not to add this clause in two places.

I'm OK to leave it as is if you have a reason, not just overlooked this variant.

(I also have some ideas why we shouldn't do it at bottom of this message.)

> I am a bit concerned that the logs are getting too long, but I don't see much we can do about it, besides removing "provided" or any other supplementary word.

Sometimes long log lines a bit inconvenient, I agree here. I also agree that we can't do much here on this level of abstraction.

OTOH, we have JSON logging and it gives some flexibility, when you're investigating a problem over cluster and/or over a lot of logs. It gives ability to highlight different pieces of information, place more important fields first, include or exclude some fields to concentrate on important details. Technically, `jq` allows these actions.

I use JSON logging a lot in the new failover coordinator service. In fact, all the messages in it are constant and the information is given using a payload. Tarantool goes toward the same direction: see for example #9830 and #9925. But here, of course, this road is longer.

Sooner or later I'm going to use box.error or some similar mechanism for structural error reporting/logging. I'm not ready to this movement on an ad-hoc basic, to be honest, because it needs to think about problems reporting in the component at whole and introducing some (possibly good) practices.

The second thing that I keep in the mind around logs is to actually describe what is going on. In particular, #9644 highlighted one place, where our configuration automation is obscure, because it doesn't print what it is actually doing.

To sum up: I wouldn't bother much about long lines in the logs in the given context.

----

I was keeping myself from discussing exact wording around the URI reporting, but since you have started this topic, I would leave my two cents.

We have two situations, when the 'bad URI' message is shown:

* Validation error. Here the URI is definitely bad. It doesn't fit the validation criteria and can't be accepted.
* The URI is skipped by the configuration building automation.

In the seconds case, the URI is not so bad. It is correct to have such an URI in `iproto.listen`. It just not suitable for replicaset/sharding configuration building automation.

So, I would say that such an URI is 'skipped' or 'not suitable'. Something like this.

OTOH, it prevents us from using the same 'bad URI' message. We can't generate it in `uri_is_suitable_to_connect()` in this case.

But maybe, we shouldn't? I just came to an idea that the `log_prefix` may be actually `log_template`. For example,

```lua
{log_template = 'replicaset dataflow configuration: skipped URI %q: %s'}
{log_template = 'sharding configuration: skipped URI %q: %s'}
```

I understand that it is going a bit too picky. So, if such ideas are not positively vibrating in your head, let's just go ahead with the code we have now.

(Just let me know about your decision.)

Reply:
I don't really like the `log_template` idea, because it's harder to understand `:format()` without a clear template. I see a lot of potentials for tipos here. 

> So, I would say that such an URI is 'skipped' or 'not suitable'. Something like this.

Ok, lets do 'unsuitable URI' for sharding/replicaset configuration.

Reply:
Done. I also removed "provided" from both messages because it seems excessive.

Reply:
Thanks!